Holovibes/cuda_sources/otsu.cu

```diff
@@ -46,98 +50,139 @@ __global__ void bradley_threshold_kernel(const float* image,
-    float localThreshold = localMean * (1 - localThresholdFactor * globalThreshold);
-    output[y * width + x] = (image[y * width + x] > localThreshold) ? 1.0f : 0.0f;
+        float localMean = localSum / count;
+        float localThreshold = localMean * (1 - localThresholdFactor * globalThreshold);
+        output[y * width + x] = (image[y * width + x] > localThreshold) ? 1.0f : 0.0f;
+    }
 }
 
-float otsu_threshold(float* d_image, int size, const cudaStream_t stream)
+__global__ void otsu_threshold_kernel(uint* hist, int total, float* threshold_out)
 {
-    uint threads = get_max_threads_1d();
-    uint blocks = map_blocks_to_problem(size, threads);
-
-    // get histogram
-    int h_hist[NUM_BINS];
-    int* d_hist;
-    cudaMalloc(&d_hist, NUM_BINS * sizeof(int));
-    cudaMemset(d_hist, 0, NUM_BINS * sizeof(int));
-    histogram_kernel<<<blocks, threads, 0, stream>>>(d_image, d_hist, size);
-    cudaDeviceSynchronize();
-    cudaMemcpy(h_hist, d_hist, NUM_BINS * sizeof(int), cudaMemcpyDeviceToHost);
-    cudaFree(d_hist);
-
-    // Compute optimal threshold
-    int total = size;
-    float sum = 0, sumB = 0, varMax = 0;
+    __shared__ float sum_shared;
+    __shared__ float varMax_shared;
+    __shared__ float threshold_shared;
+
+    int tid = threadIdx.x;
+    if (tid == 0)
+    {
+        sum_shared = 0;
+        varMax_shared = 0;
+        threshold_shared = 0;
+    }
+    __syncthreads();
+
+    // Compute total sum in parallel
+    __shared__ float partial_sum[NUM_BINS];
+    partial_sum[tid] = (tid < NUM_BINS) ? tid * hist[tid] : 0;
+    __syncthreads();
+
+    // Reduce to get total sum
+    for (int offset = blockDim.x / 2; offset > 0; offset >>= 1)
+    {
+        if (tid < offset)
+            partial_sum[tid] += partial_sum[tid + offset];
+        __syncthreads();
+    }
+
+    if (tid == 0)
+        sum_shared = partial_sum[0];
+    __syncthreads();
+
+    // Variables for Otsu
     int wB = 0, wF = 0;
-    float threshold = 0;
+    float sumB = 0;
+    float total_sum = sum_shared;
 
-    for (int i = 0; i < NUM_BINS; i++)
-        sum += i * h_hist[i];
-    for (int t = 0; t < NUM_BINS; t++)
+    for (int t = tid; t < NUM_BINS; t += blockDim.x)
     {
-        wB += h_hist[t];
+        wB += hist[t];
         if (wB == 0)
             continue;
+
         wF = total - wB;
         if (wF == 0)
             break;
 
-        sumB += t * h_hist[t];
+        sumB += t * hist[t];
         float mB = sumB / wB;
-        float mF = (sum - sumB) / wF;
+        float mF = (total_sum - sumB) / wF;
         float varBetween = wB * wF * (mB - mF) * (mB - mF);
 
-        if (varBetween > varMax)
-        {
-            varMax = varBetween;
-            threshold = t;
-        }
+        atomicMax(reinterpret_cast<unsigned int*>(&varMax_shared), __float_as_uint(varBetween));
+
+        if (varBetween == varMax_shared)
+            atomicExch(reinterpret_cast<unsigned int*>(&threshold_shared), t);
```

Comment:
You could add a little bit of documentation to explain what is the process and all the utilities of those variables. Because the names are not very meaningful.